include/revng/ABI/FunctionType.h

```diff
@@ -0,0 +1,94 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "revng/Model/Types.h"
+
+namespace abi::FunctionType {
+
+/// Best effort `CABIFunctionType` to `RawFunctionType` conversion.
+///
+/// If `ABI` is not specified, `TheBinaryToAddSupportTypesTo.DefaultABI`
+/// is used instead.
+std::optional<model::CABIFunctionType>
+tryConvertToCABI(const model::RawFunctionType &Function,
+                 model::Binary &TheBinaryToAddSupportTypesTo,
+                 std::optional<model::ABI::Values> ABI = std::nullopt);
+
+/// Best effort `RawFunctionType` to `CABIFunctionType` conversion.
+///
+/// \note: this convertion is loosy since there's no way to represent some types
+/// in `RawFunctionType` in a reversible manner.
+std::optional<model::RawFunctionType>
+tryConvertToRaw(const model::CABIFunctionType &Function,
+                model::Binary &TheBinaryToAddSupportTypesTo);
```

Comment:
Unlike `tryConvertToCABI`, I'd expect this to always succeed. In what situation it doesn't?
Maybe we need to turn that into an assertion.

Reply:
1. An ABI doesn't provide a register for return values but function returns something (can be an assert)
2. `model::type::size()` returns `std::nullopt` (also an easy assert)
3. The result of `distributeReturnValue` call providing number of registers insufficient to fit whatever type this function returs (pretty much, when `distributeReturnValue` fails in an unclear manner), (I guess this can also be an assert, since it should never happen, save for some bug with the distribution)
4. Some potential issues with stack conversions that are not implemented yet (?)

Except for the forth one I see no opposition to making this function return a value.

Reply:
1. Ah, but in that case the problem is not the conversion to `RawFunctionType`, the problems is that the `CABIFunctionType` contains information incompatible with its `ABI` field. If you want to check this (if that ever happens), it's `CABIFunctionType::verify` business.
2. Under the assumption (which can make) that the input is verifies, you can assert that.
3. `assert` FTW.
4. If there will be such failures, that will call for a review of our design. As for now, I think we can assume going to raw never fails, just as `layout` should never fail (they serve similar purpose).

Reply:
Sounds good, one less `optional`, ya-a-ay :D

Reply:
Done: https://github.com/revng/revng/pull/131/commits/45ef15cef4660f04950930cfcb50171079ac55b5